File: test-case/nest/0341/nestedint.cpp
#ifdef __clang__
    #include "soda/clang/stdc++.h"
#else
    #include <bits/stdc++.h>
#endif
#include "soda/leetcode/leetcode.h"
#include "soda/unittest/unittest.h"

using namespace std;
using namespace soda::leetcode;
using namespace soda::unittest;

// step [1]: implement class Solution
// class Solution {};
class NestedIterator {
    struct Node {
        const vector<NestedInteger>& List;
        int index;
        Node(const vector<NestedInteger>& List)
            : List{List}, index{0}
        {}
        bool isEnd() const { return index >= int(List.size()); }
        int value() const { return current().getInteger(); }
        const NestedInteger& current() const {
            return List[index];
        }
        void advance() {
            ++index;
        }
    };
public:
    NestedIterator(vector<NestedInteger> &nestedList): stk{} {
        stk.push(new Node(nestedList));
        locate();
    }
    
    int next() {
        int value = stk.top()->value();
        stk.top()->advance();
        locate();
        return value;
    }
    
    bool hasNext() {
        return !stk.empty() && !stk.top()->isEnd();
    }
private:
    stack<Node*> stk;

    void locate() {
        while (!stk.empty()) {
            if (stk.top()->isEnd()) {
                delete stk.top();
                stk.pop();
                if (!stk.empty()) {
                    stk.top()->advance();
                }
            } else if (stk.top()->current().isInteger()) {
                break;
            } else {
                stk.push(new Node(stk.top()->current().getList()));
            }
        }
    }
};

namespace {
    const auto __ = []() {
        ios_base::sync_with_stdio(false);
        cin.tie(nullptr);
        return 0;
    }();
}

class Solution {
public:
    vector<int> flatNested(vector<NestedInteger>& niList) {
        vector<int> res;
        NestedIterator iter{niList};
        while (iter.hasNext()) {
            int val = iter.next();
            res.push_back(val);
        }
        return res;
    }

    void dump(const NestedInteger& i) {
        if (i.isInteger()) {
            cerr << i.getInteger() << ' ';
        } else {
            for (auto& j : i.getList()) {
                dump(j);
            }
        }
    }
};

// [optional] use custom type with parser and serializer,
// and if you use this macro, may be you should define your own validator
// USE_CUSTOM_SERIALIZER(type)

// [optional] instantiate json access type. This will increase compile time
// #include "soda/unittest/json_access.h"
// SODA_JSON_ACCESS_TYPE(type)

int main()
{
    // [1] create by class member function
    Solution su;
    auto work = WorkFactory::create(su, &Solution::flatNested);
    //
    // [2] or, create by ordinary function
    // auto work = WorkFactory::create(function);

    // work->setValidator(validate);
    work->setCompareSerial(true);
    // work->setArgParser<0,from_type>(parse_func);
    // work->setResultParser<from_type>(parse_func);
    // work->setResultSerializer(serial_func);
    work->run();
    delete work;
    return 0;
}

File: test-case/nest/0341/Nestedint.cs
using Soda.Unittest;
using Soda.Leetcode;


public class Solution {
    public IList<int> FlatNested(IList<NestedInteger> niList)
    {
        var res = new List<int>();
        var iter = new NestedIterator(niList);
        while (iter.HasNext()) {
            res.Add(iter.Next());
        }
        return res;
    }
}

public class NestedIterator {

    class Node {
        public IList<NestedInteger> list;
        public int index;
        public Node(IList<NestedInteger> list) {
            this.list = list;
        }
        public bool IsEnd() { return index >= list.Count; }
        public int Value() { return Current().GetInteger(); }
        public NestedInteger Current() { return list[index]; }
    }

    private readonly Stack<Node> stk = new Stack<Node>();

    public NestedIterator(IList<NestedInteger> nestedList) {
        stk.Push(new Node(nestedList));
        locate();
    }

    private void locate() {
        while (stk.Count > 0) {
            if (stk.Peek().IsEnd()) {
                stk.Pop();
                if (stk.Count > 0) {
                    ++stk.Peek().index;
                }
            } else if (stk.Peek().Current().IsInteger()) {
                break;
            } else {
                stk.Push(new Node(stk.Peek().Current().GetList()));
            }
        }
    }

    public bool HasNext() {
        return stk.Count > 0 && !stk.Peek().IsEnd();
    }

    public int Next() {
        int value = stk.Peek().Value();
        ++stk.Peek().index;
        locate();
        return value;
    }
}

public class Nestedint
{
    public static void Main(string[] args)
    {
        var work = WorkFactory.Create(Utils.Fn(new Solution().FlatNested));
        // work.SetValidator((e, r) => ...);
        work.CompareSerial = true;
        Console.WriteLine(work.Run(Utils.readStdin()));
    }
}
